lib/SILOptimizer/Mandatory/AddressLowering.cpp

```diff
@@ -1248,36 +1302,60 @@ bool OpaqueStorageAllocation::findProjectionIntoUseImpl(
-    // Get the user's value, whose storage we will project into.
+    // Get the user's value, whose storage we would project into.
     SILValue userValue = getProjectedUseValue(use);
     if (!userValue)
       continue;
 
-    assert(!getProjectedDefOperand(userValue)
-           && "storage cannot project in two directions.");
+    assert(!getProjectedDefOperand(userValue) &&
+           "opcode that is both a use projection and def projection!?");
 
     // Avoid handling preposterous types.
     if (use->getOperandNumber() > UINT16_MAX)
       continue;
 
-    // Recurse through all storage projections to find the uniquely allocated
-    // storage. Enum storage cannot be reused across multiple subobjects because
-    // it must be initialized via a single init_enum_data_addr instruction.
-    //
-    // TODO: fix the memory verifier to consider the actual store instructions
-    // to initialize an enum rather than the init_enum_data_addr to reuse enum
-    // storage across multiple subobjects within the payload.
-    auto *baseStorage =
-        pass.valueStorageMap.getBaseStorage(userValue, /*allowInit*/ !intoPhi);
-    if (!baseStorage)
+    // If the user is not a phi (`intoPhi` == false), then it is always*
+    // possible to materialize initialization at the single point at which the
+    // address must be available.  *Subject to the following dominance check.
+    if (intoPhi &&
+        llvm::any_of(pass.valueStorageMap.getProjections(userValue),
+                     [&](auto *pair) { return pair->storage.initializes; })) {
+      // Materializing an address for a coalesced phi (`intoPhi` == true),
+      // however, cannot rematerialize initialization, because that would
+      // require the address to be available on both sides of the phi.  But we
+      // can't create an address phi.
+      //
+      // Concretely, given:
+      //
+      //     left:
+      //       %e1 = init_existential_value %v1
+      //       br merge(%e1 : $P)
+      //     right:
+      //       %e2 = init_existential_value %v2
+      //       br merge(%e2 : $P)
+      //     merge(%e : @owned $P):
+      //
+      // we can't produce a single init_existential_addr instruction in the
+      // `merge` block
+      //
+      //     merge:
+      //       init_existential_addr ???
+      //
+      // because doing so would require an address phi
+      //
+      //     merge(%addr : $*): // invalid!
+      //       init_existential_addr %addr
       continue;
+    }
 
-    if (auto *stackInst =
-            dyn_cast<AllocStackInst>(baseStorage->storageAddress)) {
-      if (!checkStorageDominates(stackInst, incomingValues))
-        continue;
-    } else
-      assert(isa<SILFunctionArgument>(baseStorage->storageAddress));
+    // Recurse through all storage projections to find the point where the
+    // storage has been allocated and any opening instructions involved in
+    // any of those projections' types.
```

Comment:
This is all good.

But it may be more clear if we also outright say something like:

```
The base storage address and all of the opened types used by the projections must dominate 'incomingValues' because we need to materialize corresponding address projections at those points.
```

Reply:
Done.